From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alex Henning <elcron@gmail.com>
Date: Wed, 13 Aug 2014 18:22:40 -0400
Subject: [PATCH 0294/6262] Increased build speed + caching of debs.

Make sure to add -j to improve build time by 10 minutes or more.

Change-Id: If2838a4042592fbfd1a0f40882efd7dd70270266
---
 simulation/debs/Makefile | 164 +++++++++++++++++++++++----------------
 1 file changed, 99 insertions(+), 65 deletions(-)

diff --git a/simulation/debs/Makefile b/simulation/debs/Makefile
index e526d7bf145b1233d14c6ff91de55c87526ba0f8..8091258089237f8738492ee51e5977bfd3c6ca75 100644
--- a/simulation/debs/Makefile
+++ b/simulation/debs/Makefile
@@ -16,34 +16,21 @@ libwpilibsim-package-version = $(libwpilibsim-version)-$(package-version)
 frcsim-version = 0.1
 frcsim-package-version = $(frcsim-version)-$(package-version)
 
+# Main Targets
+.PHONY: all jenkins allwpilib pull debs update-repository clean-repository clean install
+
 all: debs update-repository
 
-jenkins: download-models pull-eclipse-toolchain pull all
+jenkins: orig debs update-repository
 
 allwpilib:
 	cd $(allwpilib) && mvn -T 8 clean package -Dwith-eclipse-plugins -DskipTests -DskipIT
 
-orig: pre-orig-clean
-	cd frcsim-gazebo-plugins && tar --exclude="./debian" -czvf \
-		frcsim-gazebo-plugins_${gazebo-plugins-version}.orig.tar.gz frcsim-gazebo-plugins
-	cd frcsim-gazebo-models && tar --exclude="./debian" -czvf \
-		frcsim-gazebo-models_${gazebo-models-version}.orig.tar.gz frcsim-gazebo-models
-	cd frcsim-eclipse-plugins && tar --exclude="./debian" -czvf \
-		frcsim-eclipse-plugins_${eclipse-plugins-version}.orig.tar.gz frcsim-eclipse-plugins
-	cd frcsim-eclipse-toolchain-plugin && tar --exclude="./debian" -czvf \
-		frcsim-eclipse-toolchain-plugin_${eclipse-toolchain-version}.orig.tar.gz frcsim-eclipse-toolchain-plugin
-	cd frcsim-libwpilibsim-cpp && tar --exclude="./debian" -czvf \
-		frcsim-libwpilibsim-cpp_${libwpilibsim-version}.orig.tar.gz frcsim-libwpilibsim-cpp
-	cd frcsim && tar --exclude="./debian" -czvf \
-		frcsim_${frcsim-version}.orig.tar.gz frcsim
+orig: -orig-frcsim-gazebo-plugins -orig-frcsim-gazebo-models -orig-frcsim-eclipse-plugins \
+	  -orig-frcsim-eclipse-toolchain-plugin -orig-frcsim-libwpilibsim-cpp -orig-frcsim
 
-debs:
-	cd frcsim-gazebo-plugins/frcsim-gazebo-plugins && debuild -us -uc -iamd64
-	cd frcsim-gazebo-models/frcsim-gazebo-models && debuild -us -uc
-	cd frcsim-eclipse-plugins/frcsim-eclipse-plugins && debuild -us -uc
-	cd frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin && debuild -us -uc
-	cd frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp && debuild -us -uc
-	cd frcsim/frcsim && debuild -us -uc
+debs: -debs-frcsim-gazebo-plugins -debs-frcsim-gazebo-models -debs-frcsim-eclipse-plugins \
+	  -debs-frcsim-eclipse-toolchain-plugin -debs-frcsim-libwpilibsim-cpp -debs-frcsim
 
 update-repository: clean-repository
 	cd repository && reprepro includedeb $(codename) ../frcsim-gazebo-plugins/frcsim-gazebo-plugins_$(gazebo-plugins-package-version)_amd64.deb
@@ -61,40 +48,52 @@ clean-repository:
 	cd repository && reprepro remove $(codename) frcsim-libwpilibsim-cpp
 	cd repository && reprepro remove $(codename) frcsim
 
-pre-orig-clean:
-	cd frcsim-gazebo-plugins/frcsim-gazebo-plugins && debuild clean
-	cd frcsim-gazebo-models/frcsim-gazebo-models && debuild clean
-	cd frcsim-eclipse-plugins/frcsim-eclipse-plugins && debuild clean
-	cd frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin && debuild clean
-	cd frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp && debuild clean
-	cd frcsim/frcsim && debuild clean
-
-	rm -f frcsim-gazebo-plugins/frcsim-gazebo-plugins_$(gazebo-plugins-version).orig.tar.gz
-	rm -f frcsim-gazebo-models/frcsim-gazebo-models_$(gazebo-models-version).orig.tar.gz
-	rm -f frcsim-eclipse-plugins/frcsim-eclipse-plugins_$(eclipse-plugins-version).orig.tar.gz
-	rm -f frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin_$(eclipse-toolchain-version).orig.tar.gz
-	rm -f frcsim-libwpilib-cpp/frcsim-libwpilib-cpp_$(libwpilibsim-version).orig.tar.gz
-	rm -f frcsim/frcsim_$(frcsim-version).orig.tar.gz
-
-clean: pre-orig-clean
-	rm -rf frcsim-eclipse-plugins/frcsim-eclipse-plugins/plugins
-	rm -rf frcsim-eclipse-plugins/frcsim-eclipse-plugins/features
-
-	rm -rf frcsim-gazebo-plugins/frcsim-gazebo-plugins/src
-
-	rm -rf frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp/src
-	rm -rf frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp/include
-	rm -f frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp/Makefile
-	rm -f frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp/CMakeLists.txt
-	rm -rf frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp/target
+clean:
+	-rm frcsim-gazebo-models/frcsim-gazebo-models_${gazebo-models-version}.orig.tar.gz
+	-rm frcsim-gazebo-plugins/frcsim-gazebo-plugins_${gazebo-plugins-version}.orig.tar.gz
+	-rm frcsim-eclipse-plugins/frcsim-eclipse-plugins_${eclipse-plugins-version}.orig.tar.gz
+	-rm frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin_${eclipse-toolchain-version}.orig.tar.gz
+	-rm frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp_${libwpilibsim-version}.orig.tar.gz
+	-rm frcsim/frcsim_$(frcsim-package-version)_all.deb
+
+	-rm frcsim-gazebo-plugins/frcsim-gazebo-plugins_$(gazebo-plugins-package-version)_amd64.deb
+	-rm frcsim-gazebo-models/frcsim-gazebo-models_$(gazebo-models-package-version)_all.deb
+	-rm frcsim-eclipse-plugins/frcsim-eclipse-plugins_$(eclipse-plugins-package-version)_all.deb
+	-rm frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin_$(eclipse-toolchain-package-version)_all.deb
+	-rm frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp_$(libwpilibsim-package-version)_amd64.deb
+	-rm frcsim/frcsim_$(frcsim-package-version)_all.deb
 
-pull: clean pull-gazebo-plugins pull-eclipse-plugins pull-libwpilibsim-cpp orig
+install:
+	sudo dpkg -i frcsim-gazebo-plugins/frcsim-gazebo-plugins_$(gazebo-plugins-package-version)_amd64.deb \
+	frcsim-gazebo-models/frcsim-gazebo-models_$(gazebo-models-package-version)_all.deb \
+	frcsim-eclipse-plugins/frcsim-eclipse-plugins_$(eclipse-plugins-package-version)_all.deb \
+	frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin_$(eclipse-toolchain-package-version)_all.deb \
+	frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp_$(libwpilibsim-package-version)_amd64.deb \
+	frcsim/frcsim_$(frcsim-package-version)_all.deb
+
+# orig support (for faster parallel builds)
+.PHONY: -orig-frcsim-gazebo-plugins -orig-frcsim-gazebo-models -orig-frcsim-eclipse-plugins \
+	  	-orig-frcsim-eclipse-toolchain-plugin -orig-frcsim-libwpilibsim-cpp -orig-frcsim
+.PHONY:	-pull-gazebo-plugins -pull-eclipse-plugins -pull-libwpilibsim-cpp
+
+-orig-frcsim-gazebo-models: frcsim-gazebo-models/frcsim-gazebo-models_${gazebo-models-version}.orig.tar.gz
+frcsim-gazebo-models/frcsim-gazebo-models_${gazebo-models-version}.orig.tar.gz:
+	cd frcsim-gazebo-models && \
+	    wget --no-check-certificate $(gazebo-models-orig-url) -O frcsim-gazebo-models_${gazebo-models-version}.orig.tar.gz && \
+	    tar xvf frcsim-gazebo-models_${gazebo-models-version}.orig.tar.gz
 
-pull-gazebo-plugins:
+-orig-frcsim-gazebo-plugins: frcsim-gazebo-plugins/frcsim-gazebo-plugins_${gazebo-plugins-version}.orig.tar.gz
+-pull-gazebo-plugins:
 	cp -rf -t frcsim-gazebo-plugins/frcsim-gazebo-plugins/ $(allwpilib)/simulation/frc_gazebo_plugins/*
 	echo Increment version?
+frcsim-gazebo-plugins/frcsim-gazebo-plugins_${gazebo-plugins-version}.orig.tar.gz: -pull-gazebo-plugins
+	cd frcsim-gazebo-plugins/frcsim-gazebo-plugins && debuild clean
+	rm -f frcsim-gazebo-plugins/frcsim-gazebo-plugins_$(gazebo-plugins-version).orig.tar.gz
+	cd frcsim-gazebo-plugins && tar --exclude="./debian" -czvf \
+		frcsim-gazebo-plugins_${gazebo-plugins-version}.orig.tar.gz frcsim-gazebo-plugins
 
-pull-eclipse-plugins:
+-orig-frcsim-eclipse-plugins: frcsim-eclipse-plugins/frcsim-eclipse-plugins_${eclipse-plugins-version}.orig.tar.gz
+-pull-eclipse-plugins:
 	rm -rf frcsim-eclipse-plugins/frcsim-eclipse-plugins/plugins
 	rm -rf frcsim-eclipse-plugins/frcsim-eclipse-plugins/features
 	mkdir -p frcsim-eclipse-plugins/frcsim-eclipse-plugins/plugins
@@ -106,8 +105,14 @@ pull-eclipse-plugins:
 	cp $(allwpilib)/eclipse-plugins/edu.wpi.first.wpilib.plugins.updatesite/target/site/features/edu.wpi.first.wpilib.plugins.cpp.feature_* frcsim-eclipse-plugins/frcsim-eclipse-plugins/features
 	cp $(allwpilib)/eclipse-plugins/edu.wpi.first.wpilib.plugins.updatesite/target/site/features/edu.wpi.first.wpilib.plugins.core.feature_* frcsim-eclipse-plugins/frcsim-eclipse-plugins/features
 	echo Increment version?
+frcsim-eclipse-plugins/frcsim-eclipse-plugins_${eclipse-plugins-version}.orig.tar.gz: -pull-eclipse-plugins
+	cd frcsim-eclipse-plugins/frcsim-eclipse-plugins && debuild clean
+	rm -f frcsim-eclipse-plugins/frcsim-eclipse-plugins_$(eclipse-plugins-version).orig.tar.gz
+	cd frcsim-eclipse-plugins && tar --exclude="./debian" -czvf \
+		frcsim-eclipse-plugins_${eclipse-plugins-version}.orig.tar.gz frcsim-eclipse-plugins
 
-pull-eclipse-toolchain:
+-orig-frcsim-eclipse-toolchain-plugin: frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin_${eclipse-toolchain-version}.orig.tar.gz
+-pull-eclipse-toolchain:
 	rm -rf frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin/plugins
 	rm -rf frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin/features
 	mkdir -p frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin/plugins
@@ -115,24 +120,53 @@ pull-eclipse-toolchain:
 	cp $(allwpilib)/eclipse-plugins/edu.wpi.first.wpilib.plugins.updatesite/target/site/plugins/edu.wpi.first.wpilib.plugins.cpp.toolchains.linux_* frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin/plugins/
 	cp $(allwpilib)/eclipse-plugins/edu.wpi.first.wpilib.plugins.updatesite/target/site/features/edu.wpi.first.wpilib.plugins.cpp.toolchains.linux.feature_* frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin/features/
 	echo Increment version?
+frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin_${eclipse-toolchain-version}.orig.tar.gz: -pull-eclipse-toolchain
+	cd frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin && debuild clean
+	rm -f frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin_$(eclipse-toolchain-version).orig.tar.gz
+	cd frcsim-eclipse-toolchain-plugin && tar --exclude="./debian" -czvf \
+		frcsim-eclipse-toolchain-plugin_${eclipse-toolchain-version}.orig.tar.gz frcsim-eclipse-toolchain-plugin
 
-pull-libwpilibsim-cpp:
+-orig-frcsim-libwpilibsim-cpp: frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp_${libwpilibsim-version}.orig.tar.gz
+-pull-libwpilibsim-cpp:
 	cp -rf -t frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp/ $(allwpilib)/wpilibc/wpilibC++Sim/*
 	echo Increment version?
+frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp_${libwpilibsim-version}.orig.tar.gz: -pull-libwpilibsim-cpp
+	cd frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp && debuild clean
+	rm -f frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp_$(libwpilibsim-version).orig.tar.gz
+	cd frcsim-libwpilibsim-cpp && tar --exclude="./debian" -czvf \
+		frcsim-libwpilibsim-cpp_${libwpilibsim-version}.orig.tar.gz frcsim-libwpilibsim-cpp
 
-download-models:
-	cd frcsim-gazebo-models && \
-	    wget --no-check-certificate $(gazebo-models-orig-url) -O frcsim-gazebo-models_${gazebo-models-version}.orig.tar.gz && \
-	    tar xvf frcsim-gazebo-models_${gazebo-models-version}.orig.tar.gz
+-orig-frcsim: frcsim/frcsim_$(frcsim-package-version)_all.deb
+frcsim/frcsim_${frcsim-version}.orig.tar.gz:
+	cd frcsim/frcsim && debuild clean
+	rm -f frcsim/frcsim_$(frcsim-version).orig.tar.gz
+	cd frcsim && tar --exclude="./debian" -czvf \
+		frcsim_${frcsim-version}.orig.tar.gz frcsim
 
-install:
-	sudo dpkg -i frcsim-gazebo-plugins/frcsim-gazebo-plugins_0.2-1_amd64.deb \
-                 frcsim-gazebo-models/frcsim-gazebo-models_0.2-1_all.deb \
-                 frcsim-eclipse-plugins/frcsim-eclipse-plugins_0.1-1_all.deb \
-                 frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin_0.1-1_all.deb \
-                 frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp_0.1-1_amd64.deb \
-                 frcsim/frcsim_0.1-1_all.deb
+# debs support (for faster parallel builds)
+.PHONY: -debs-frcsim-gazebo-plugins -debs-frcsim-gazebo-models -debs-frcsim-eclipse-plugins \
+		-debs-frcsim-eclipse-toolchain-plugin -debs-frcsim-libwpilibsim-cpp -debs-frcsim
+
+-debs-frcsim-gazebo-plugins: frcsim-gazebo-plugins/frcsim-gazebo-plugins_$(gazebo-plugins-package-version)_amd64.deb
+frcsim-gazebo-plugins/frcsim-gazebo-plugins_$(gazebo-plugins-package-version)_amd64.deb: frcsim-gazebo-plugins/frcsim-gazebo-plugins_$(gazebo-plugins-version).orig.tar.gz
+	cd frcsim-gazebo-plugins/frcsim-gazebo-plugins && debuild --no-lintian -us -uc -iamd64
+
+-debs-frcsim-gazebo-models: frcsim-gazebo-models/frcsim-gazebo-models_$(gazebo-models-package-version)_all.deb
+frcsim-gazebo-models/frcsim-gazebo-models_$(gazebo-models-package-version)_all.deb: frcsim-gazebo-models/frcsim-gazebo-models_$(gazebo-models-version).orig.tar.gz
+	cd frcsim-gazebo-models/frcsim-gazebo-models && debuild --no-lintian -us -uc
+
+-debs-frcsim-eclipse-plugins: frcsim-eclipse-plugins/frcsim-eclipse-plugins_$(eclipse-plugins-package-version)_all.deb
+frcsim-eclipse-plugins/frcsim-eclipse-plugins_$(eclipse-plugins-package-version)_all.deb: frcsim-eclipse-plugins/frcsim-eclipse-plugins_$(eclipse-plugins-version).orig.tar.gz
+	cd frcsim-eclipse-plugins/frcsim-eclipse-plugins && debuild --no-lintian -us -uc
+
+-debs-frcsim-eclipse-toolchain-plugin: frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin_$(eclipse-toolchain-package-version)_all.deb
+frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin_$(eclipse-toolchain-package-version)_all.deb: frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin_$(eclipse-toolchain-version).orig.tar.gz
+	cd frcsim-eclipse-toolchain-plugin/frcsim-eclipse-toolchain-plugin && debuild --no-lintian -us -uc
 
-deploy:
-	rsync -r -v -C -p repository/ adhenning@ccc.wpi.edu:public_html/frcsim/
+-debs-frcsim-libwpilibsim-cpp: frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp_$(libwpilibsim-package-version)_amd64.deb
+frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp_$(libwpilibsim-package-version)_amd64.deb: frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp_$(libwpilibsim-version).orig.tar.gz
+	cd frcsim-libwpilibsim-cpp/frcsim-libwpilibsim-cpp && debuild --no-lintian -us -uc
 
+-debs-frcsim: frcsim/frcsim_$(frcsim-package-version)_all.deb
+frcsim/frcsim_$(frcsim-package-version)_all.deb: frcsim/frcsim_$(frcsim-version).orig.tar.gz
+	cd frcsim/frcsim && debuild --no-lintian -us -uc
